File: android/app/src/main/java/app/notomorrow/feature/workout/PlateCalculatorSheet.kt
package app.notomorrow.feature.workout

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import app.notomorrow.designsystem.Chip
import app.notomorrow.designsystem.Eyebrow
import app.notomorrow.designsystem.GhostButton
import app.notomorrow.designsystem.Hairline
import app.notomorrow.designsystem.NT
import app.notomorrow.designsystem.NtSheet
import app.notomorrow.designsystem.NtShapes
import app.notomorrow.designsystem.NtText
import app.notomorrow.designsystem.SecondaryButton
import app.notomorrow.designsystem.TabularText
import app.notomorrow.designsystem.ntMediumDetent
import app.notomorrow.designsystem.ntPlainClickable
import app.notomorrow.di.LocalAppContainer
import app.notomorrow.feature.progress.OneRepMaxCalculatorSheet
import app.notomorrow.model.WeightUnit
import app.notomorrow.util.Fmt
import app.notomorrow.util.S
import kotlinx.coroutines.launch
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

/**
 * Plate calculator for one set's weight — 1:1 port of `PlateCalculatorSheet.swift` (the "Plates"
 * accessory over the keyboard in the active workout): bar choice, a drawing of the loaded bar, the
 * plates per side, and, when the plates cannot make the weight, the closest load with Use.
 *
 * [weightKg] is the set's stored weight; [onUse] replaces it (kg) with a load the plates can make.
 * The bar is remembered per unit (`nt.plates.barKg` / `nt.plates.barLb`).
 */
@Composable
fun PlateCalculatorSheet(
    weightKg: Double,
    unit: WeightUnit,
    onUse: (Double) -> Unit,
    onDismiss: () -> Unit,
) {
    val prefs = LocalAppContainer.current.appPrefs
    val scope = rememberCoroutineScope()
    val defaultBar = PlateMath.bars(unit).first()
    val bar by remember(unit) { prefs.plateBar(unit) }.collectAsState(initial = defaultBar)
    val target = SetInput.display(weightKg, unit)
    val load = PlateMath.load(target, bar, PlateMath.plates(unit), limit = PlateMath.maxTarget(unit))
    var showsOneRepMax by rememberSaveable { mutableStateOf(false) }

    NtSheet(
        onDismiss = onDismiss,
        showsHandle = true,
        containerColor = NT.Colors.ground,
        minHeight = ntMediumDetent(),
        skipPartiallyExpanded = false,
    ) {
        Header(onDismiss = onDismiss)
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = NT.Spacing.screenH)
                .padding(bottom = NT.Spacing.section),
        ) {
            TabularText(Fmt.plate(target, unit), style = NT.Fonts.display(44), color = NT.Colors.ink)
            BarPicker(
                unit = unit,
                bar = bar,
                modifier = Modifier.padding(top = 16.dp),
                onSelect = { value -> scope.launch { prefs.setPlateBar(unit, value) } },
            )
            BarbellDrawing(
                perSide = load.perSide,
                unit = unit,
                modifier = Modifier
                    .padding(top = 20.dp)
                    .fillMaxWidth()
                    .height(120.dp)
                    .clearAndSetSemantics {},
            )
            PerSideList(load = load, unit = unit, modifier = Modifier.padding(top = 20.dp))
            if (load.isOverMax) {
                Note(stringResource(S.plates_overMax_s, Fmt.plate(PlateMath.maxTarget(unit), unit)))
            } else if (load.isBelowBar) {
                Note(stringResource(S.plates_belowBar))
            } else if (!load.isExact) {
                val label = Fmt.plate(load.total, unit)
                Note(stringResource(S.plates_notExact_s, label))
                SecondaryButton(
                    title = stringResource(S.plates_use_s, label),
                    modifier = Modifier.padding(top = 12.dp),
                    onClick = {
                        onUse(SetInput.kg(load.total, unit))
                        onDismiss()
                    },
                )
            }
            GhostButton(
                title = stringResource(S.onerm_calculator),
                modifier = Modifier.padding(top = NT.Spacing.section),
                onClick = { showsOneRepMax = true },
            )
        }
    }
    if (showsOneRepMax) {
        OneRepMaxCalculatorSheet(unit = unit, onDismiss = { showsOneRepMax = false }, initialWeight = target)
    }
}

@Composable
private fun Header(onDismiss: () -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = NT.Spacing.screenH)
            .padding(top = 16.dp, bottom = 8.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        NtText(stringResource(S.plates_title), style = NT.Fonts.title2, color = NT.Colors.ink, maxLines = 1)
        Spacer(Modifier.weight(1f))
        Box(
            modifier = Modifier
                .heightIn(min = NT.Size.control)
                .ntPlainClickable(onClick = onDismiss),
            contentAlignment = Alignment.Center,
        ) {
            NtText(stringResource(S.common_done), style = NT.Fonts.body, color = NT.Colors.ink2, maxLines = 1)
        }
    }
}

@Composable
private fun BarPicker(unit: WeightUnit, bar: Double, modifier: Modifier, onSelect: (Double) -> Unit) {
    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Eyebrow(stringResource(S.plates_bar))
        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            for (value in PlateMath.bars(unit)) {
                Chip(title = Fmt.plate(value, unit), selected = value == bar, onClick = { onSelect(value) })
            }
        }
    }
}

@Composable
private fun PerSideList(load: PlateMath.Load, unit: WeightUnit, modifier: Modifier) {
    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Eyebrow(stringResource(S.plates_perSide))
        if (load.perSide.isEmpty()) {
            NtText(stringResource(S.plates_emptyBar), style = NT.Fonts.body, color = NT.Colors.ink2)
        } else {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .background(NT.Colors.surface, NtShapes.tile)
                    .padding(horizontal = 16.dp),
            ) {
                load.groups.forEachIndexed { index, group ->
                    if (index > 0) Hairline()
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .heightIn(min = NT.Size.control)
                            .semantics(mergeDescendants = true) {},
                        horizontalArrangement = Arrangement.spacedBy(8.dp),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        Box(Modifier.size(10.dp).background(PlateStyle.color(group.plate, unit), CircleShape))
                        TabularText(Fmt.plate(group.plate, unit), style = NT.Fonts.body, color = NT.Colors.ink)
                        Spacer(Modifier.weight(1f))
                        TabularText(
                            text = Fmt.TIMES + " " + group.count,
                            style = NT.Fonts.headline,
                            color = NT.Colors.ink,
                        )
                    }
                }
            }
        }
    }
}

@Composable
private fun Note(text: String) {
    NtText(
        text = text,
        modifier = Modifier.padding(top = 16.dp),
        style = NT.Fonts.footnote,
        color = NT.Colors.ink2,
    )
}

// MARK: - Drawing

/**
 * A barbell seen from the front: the sleeve on each side with its plates, heaviest nearest the
 * middle, taller for heavier plates, in the usual competition colours.
 */
@Composable
private fun BarbellDrawing(perSide: List<Double>, unit: WeightUnit, modifier: Modifier) {
    val shaft = NT.Colors.surface3
    val collar = NT.Colors.ink3
    val maxPlate = PlateMath.plates(unit).first()
    Canvas(modifier) {
        val midY = size.height / 2
        // Shaft and sleeves.
        drawRoundRect(
            color = shaft,
            topLeft = Offset(0f, midY - 4.dp.toPx()),
            size = Size(size.width, 8.dp.toPx()),
            cornerRadius = CornerRadius(4.dp.toPx()),
        )
        for (x in listOf(size.width * 0.3f, size.width * 0.7f)) {
            drawRect(
                color = collar,
                topLeft = Offset(x - 3.dp.toPx(), midY - 13.dp.toPx()),
                size = Size(6.dp.toPx(), 26.dp.toPx()),
            )
        }
        side(perSide, unit, maxPlate, leading = true)
        side(perSide, unit, maxPlate, leading = false)
    }
}

private fun DrawScope.side(perSide: List<Double>, unit: WeightUnit, maxPlate: Double, leading: Boolean) {
    val midY = size.height / 2
    val collarX = if (leading) size.width * 0.3f else size.width * 0.7f
    val room = size.width * 0.3f - 4.dp.toPx()
    val plateWidth = min(16.dp.toPx(), max(5.dp.toPx(), room / max(perSide.size, 1) - 2.dp.toPx()))
    val radius = CornerRadius(3.dp.toPx())
    perSide.forEachIndexed { index, plate ->
        val offset = 6.dp.toPx() + (plateWidth + 2.dp.toPx()) * index + plateWidth / 2
        val ratio = max(0.35, sqrt(plate / maxPlate)).toFloat()
        val height = size.height * ratio
        val centerX = if (leading) collarX - offset else collarX + offset
        val topLeft = Offset(centerX - plateWidth / 2, midY - height / 2)
        drawRoundRect(PlateStyle.color(plate, unit), topLeft, Size(plateWidth, height), radius)
        // `strokeBorder(.black.opacity(0.35), lineWidth: 1)` — drawn inside the plate's edge.
        val inset = 0.5.dp.toPx()
        drawRoundRect(
            color = Color.Black.copy(alpha = 0.35f),
            topLeft = Offset(topLeft.x + inset, topLeft.y + inset),
            size = Size(plateWidth - 2 * inset, height - 2 * inset),
            cornerRadius = radius,
            style = Stroke(width = 1.dp.toPx()),
        )
    }
}

/** Competition plate colours by weight (red 25 · blue 20 · yellow 15 · green 10 · white 5 · then small plates). */
internal object PlateStyle {
    fun color(plate: Double, unit: WeightUnit): Color {
        val kg = if (unit == WeightUnit.Kg) plate else plate / Fmt.LB_PER_KG
        return when {
            kg >= 24 -> Color(0xFFE5484D)
            kg >= 19 -> Color(0xFF3E7BFA)
            kg >= 14 -> Color(0xFFF5C04A)
            kg >= 9 -> Color(0xFF30A46C)
            kg >= 4 -> Color(0xFFEDEDED)
            kg >= 2 -> Color(0xFF6E6E73)
            else -> Color(0xFFB8B8BD)
        }
    }
}

File: NoTomorrow/Features/Workout/PlateCalculatorSheet.swift
import SwiftUI

/// Plate calculator for one set's weight (keyboard toolbar "Plates" in the active workout): bar choice, a drawing of
/// the loaded bar, the plates per side, and, when the plates cannot make the weight, the closest load with Use.
struct PlateCalculatorSheet: View {
    let weightKg: Double
    let unit: WeightUnit
    /// Replaces the set's weight (kg) with a load the plates can make.
    var onUse: (Double) -> Void

    @Environment(\.dismiss) private var dismiss
    @AppStorage("nt.plates.barKg") private var barKg: Double = 20
    @AppStorage("nt.plates.barLb") private var barLb: Double = 45
    @State private var showsOneRepMax = false

    private var target: Double { SetInput.display(weightKg, unit: unit) }
    private var bar: Double { unit == .kg ? barKg : barLb }
    private var load: PlateMath.Load { PlateMath.load(target: target, bar: bar, plates: PlateMath.plates(for: unit),
                                                                  limit: PlateMath.maxTarget(for: unit)) }

    var body: some View {
        let load = load
        VStack(alignment: .leading, spacing: 0) {
            header
            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    Text(Fmt.plate(target, unit: unit))
                        .font(NT.Fonts.display(44)).foregroundStyle(NT.Colors.ink).tabular()
                    barPicker.padding(.top, 16)
                    BarbellDrawing(perSide: load.perSide, unit: unit)
                        .frame(height: 120)
                        .padding(.top, 20)
                        .accessibilityHidden(true)
                    perSideList(load).padding(.top, 20)
                    if load.isOverMax {
                        note("plates.overMax \(Fmt.plate(PlateMath.maxTarget(for: unit), unit: unit))")
                    } else if load.isBelowBar {
                        note("plates.belowBar")
                    } else if !load.isExact {
                        closest(load)
                    }
                    GhostButton(title: "onerm.calculator", systemImage: "percent") { showsOneRepMax = true }
                        .padding(.top, NT.Spacing.section)
                }
                .padding(.horizontal, NT.Spacing.screenH)
                .padding(.bottom, NT.Spacing.section)
            }
        }
        .ntScreenBackground()
        .presentationBackground(NT.Colors.ground)
        .presentationDragIndicator(.visible)
        .presentationDetents([.medium, .large])
        .sheet(isPresented: $showsOneRepMax) { OneRepMaxCalculatorSheet(unit: unit, initialWeight: target) }
    }

    private var header: some View {
        HStack {
            Text("plates.title").font(NT.Fonts.title2).foregroundStyle(NT.Colors.ink)
            Spacer()
            Button { dismiss() } label: {
                Text("common.done").font(NT.Fonts.body).foregroundStyle(NT.Colors.ink2)
                    .frame(minHeight: NT.Size.control)
            }
        }
        .padding(.horizontal, NT.Spacing.screenH)
        .padding(.top, 16)
        .padding(.bottom, 8)
    }

    private var barPicker: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("plates.bar").eyebrow()
            HStack(spacing: 8) {
                ForEach(PlateMath.bars(for: unit), id: \.self) { value in
                    Chip(title: Fmt.plate(value, unit: unit), isSelected: value == bar) {
                        if unit == .kg { barKg = value } else { barLb = value }
                    }
                }
            }
        }
    }

    private func perSideList(_ load: PlateMath.Load) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("plates.perSide").eyebrow()
            if load.perSide.isEmpty {
                Text("plates.emptyBar").font(NT.Fonts.body).foregroundStyle(NT.Colors.ink2)
            } else {
                VStack(spacing: 0) {
                    ForEach(Array(load.groups.enumerated()), id: \.offset) { index, group in
                        if index > 0 { Hairline() }
                        HStack {
                            Circle().fill(PlateStyle.color(group.plate, unit: unit)).frame(width: 10, height: 10)
                                .accessibilityHidden(true)
                            Text(Fmt.plate(group.plate, unit: unit))
                                .font(NT.Fonts.body).foregroundStyle(NT.Colors.ink).tabular()
                            Spacer()
                            Text(verbatim: "× \(group.count)")
                                .font(NT.Fonts.headline).foregroundStyle(NT.Colors.ink).tabular()
                        }
                        .frame(minHeight: NT.Size.control)
                        .accessibilityElement(children: .combine)
                    }
                }
                .padding(.horizontal, 16)
                .background(NT.Colors.surface, in: RoundedRectangle(cornerRadius: NT.Radius.tile, style: .continuous))
            }
        }
    }

    private func closest(_ load: PlateMath.Load) -> some View {
        let label = Fmt.plate(load.total, unit: unit)
        return VStack(alignment: .leading, spacing: 12) {
            note("plates.notExact \(label)")
            SecondaryButton(title: "plates.use \(label)") {
                onUse(SetInput.kg(fromDisplay: load.total, unit: unit))
                dismiss()
            }
        }
    }

    private func note(_ key: LocalizedStringKey) -> some View {
        Text(key)
            .font(NT.Fonts.footnote).foregroundStyle(NT.Colors.ink2)
            .fixedSize(horizontal: false, vertical: true)
            .padding(.top, 16)
    }
}

// MARK: - Drawing

/// A barbell seen from the front: the sleeve on each side with its plates, heaviest nearest the middle,
/// taller for heavier plates, in the usual competition colours.
private struct BarbellDrawing: View {
    let perSide: [Double]
    let unit: WeightUnit

    var body: some View {
        GeometryReader { geo in
            let midY = geo.size.height / 2
            let maxPlate = PlateMath.plates(for: unit).first ?? 1
            ZStack {
                // Shaft and sleeves.
                Capsule().fill(NT.Colors.surface3)
                    .frame(width: geo.size.width, height: 8)
                    .position(x: geo.size.width / 2, y: midY)
                Rectangle().fill(NT.Colors.ink3)
                    .frame(width: 6, height: 26)
                    .position(x: geo.size.width * 0.3, y: midY)
                Rectangle().fill(NT.Colors.ink3)
                    .frame(width: 6, height: 26)
                    .position(x: geo.size.width * 0.7, y: midY)
                side(width: geo.size.width, midY: midY, height: geo.size.height, maxPlate: maxPlate, leading: true)
                side(width: geo.size.width, midY: midY, height: geo.size.height, maxPlate: maxPlate, leading: false)
            }
        }
    }

    private func side(width: CGFloat, midY: CGFloat, height: CGFloat, maxPlate: Double, leading: Bool) -> some View {
        let collarX = leading ? width * 0.3 : width * 0.7
        let room = width * 0.3 - 4
        let plateWidth = min(16, max(5, room / CGFloat(max(perSide.count, 1)) - 2))
        return ForEach(Array(perSide.enumerated()), id: \.offset) { index, plate in
            let offset = 6 + (plateWidth + 2) * CGFloat(index) + plateWidth / 2
            let ratio = CGFloat(max(0.35, sqrt(plate / maxPlate)))
            RoundedRectangle(cornerRadius: 3, style: .continuous)
                .fill(PlateStyle.color(plate, unit: unit))
                .overlay(RoundedRectangle(cornerRadius: 3, style: .continuous).strokeBorder(.black.opacity(0.35), lineWidth: 1))
                .frame(width: plateWidth, height: height * ratio)
                .position(x: leading ? collarX - offset : collarX + offset, y: midY)
        }
    }
}

/// Competition plate colours by weight (red 25 · blue 20 · yellow 15 · green 10 · white 5 · then small plates).
enum PlateStyle {
    static func color(_ plate: Double, unit: WeightUnit) -> Color {
        let kg = unit == .kg ? plate : plate / Fmt.lbPerKg
        switch kg {
        case 24...: return Color(hex: 0xE5484D)
        case 19..<24: return Color(hex: 0x3E7BFA)
        case 14..<19: return Color(hex: 0xF5C04A)
        case 9..<14: return Color(hex: 0x30A46C)
        case 4..<9: return Color(hex: 0xEDEDED)
        case 2..<4: return Color(hex: 0x6E6E73)
        default: return Color(hex: 0xB8B8BD)
        }
    }
}
